keycloak/openid_client.go

```diff
@@ -1,37 +1,329 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) DeleteOpenidClientAuthorizationPermission(realmId, resourceServerId, permissionId string) error {
+	return keycloakClient.delete(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/permission/%s", realmId, resourceServerId, permissionId), nil)
+}
+
+func (keycloakClient *KeycloakClient) GetClientRoleByName(realm, clientId, name string) (*OpenidClientRole, error) {
+	var clientRole OpenidClientRole
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/clients/%s/roles/%s", realm, clientId, name), &clientRole, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &clientRole, nil
+}
+
+func (keycloakClient *KeycloakClient) GetClientByName(realm, clientId string) (*OpenidClient, error) {
+	var clients []OpenidClient
+	params := map[string]string{"clientId": clientId}
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/clients", realm), &clients, params)
+	if err != nil {
+		return nil, err
+	}
+	return &clients[0], nil
+}
+
+func (keycloakClient *KeycloakClient) NewOpenidClientServiceAccountRole(serviceAccountRole *OpenidClientServiceAccountRole) error {
+	role, err := keycloakClient.GetClientRoleByName(serviceAccountRole.RealmId, serviceAccountRole.ContainerId, serviceAccountRole.Name)
+	if err != nil {
+		return err
+	}
+	serviceAccountRole.Id = role.Id
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, *serviceAccountRole)
+	_, _, err = keycloakClient.post(fmt.Sprintf("/realms/%s/users/%s/role-mappings/clients/%s", serviceAccountRole.RealmId, serviceAccountRole.ServiceAccountUserId, serviceAccountRole.ContainerId), serviceAccountRoles)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) DeleteOpenidClientServiceAccountRole(realm, serviceAccountUserId, clientId, roleId string) error {
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, OpenidClientServiceAccountRole{
+		Id: roleId,
+	})
+	err := keycloakClient.delete(fmt.Sprintf("/realms/%s/users/%s/role-mappings/clients/%s", realm, serviceAccountUserId, clientId), &serviceAccountRoles)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) GetOpenidClientServiceAccountRole(realm, serviceAccountUserId, clientId, roleId string) (*OpenidClientServiceAccountRole, error) {
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, OpenidClientServiceAccountRole{
+		Id:                   roleId,
+		RealmId:              realm,
+		ContainerId:          clientId,
+		ServiceAccountUserId: serviceAccountUserId,
+	})
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/users/%s/role-mappings/clients/%s", realm, serviceAccountUserId, clientId), &serviceAccountRoles, nil)
```

Comment:
I'm not sure I understand what's going on here.  It looks like you're initializing the `serviceAccountRoles` slice with a single element, but then it's immediately being overwritten by this line.  Is there a reason for this?

Reply:
Fixed